### PR TITLE
[istio] add a sidecar proxy injection template with preStop hook

### PR DESCRIPTION
**Pull request by fl64**
## Description
Added istio sidecar template to prevent istio-proxy from terminating before the main application's connections are closed

## Why do we need it, and what problem does it solve?
Closes https://github.com/deckhouse/deckhouse/issues/2401

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: istio
type: chore
summary: Added an istio sidecar template to prevent istio-proxy from terminating before the main application's network sockets are closed.
impact_level: default
```

### CODE REVIEW

File: ee/modules/110-istio/docs/USAGE.md [hunk 1]
```diff
@@ -473,6 +473,15 @@ spec:
...
 
+## Control the data-plane behavior
+
+### [experimental feature] Prevent istio-proxy from terminating before the main application's connections are closed
+
+By default, isiot-proxy terminates after receiving a SIGTERM signal before terminating established connections and closing sockets open for listening by the main application.
+This behavior can be corrected by setting the following annotation on the application's pod: `inject.istio.io/templates: sidecar,d8-hold-istio-proxy-termination-until-application-stops`.
+Setting this annotation will force sidecar-injector to insert a preStop-hook for istio-proxy, which when receiving the SIGTERM signal will cause istio-proxy to wait for all established connections to be closed and sockets open for listening, but no more than the `terminationGracePeriodSeconds` interval.
+In some cases, this will prevent the application from terminating incorrectly.
+
```
Comment (apolovov) [inline ee/modules/110-istio/docs/USAGE.md:484]:
```suggestion
By default, during termination, all containers in a Pod, including istio-proxy one, receive SIGTERM signal simultanuesly. But some applications need time to properly handle the termination and sometimes they need to do some network requests. It isn't possible when the istio-proxy stops before the application do. The solution is to add a preStop hook which evaluates the application's network sockets status and stops when they aren't in network namespace.
The annotation below adds the preStop hook to istio-proxy container in application's Pod:
`inject.istio.io/templates: sidecar,d8-hold-istio-proxy-termination-until-application-stops`.
```

File: ee/modules/110-istio/docs/USAGE_RU.md [hunk 2]
```diff
@@ -473,6 +473,15 @@ spec:
...
 ```
 
+## Управление поведением data-plane
+
+### [экспериментальная функция] Предотвратить завершение работы istio-proxy до завершения соединений основного приложения
+
+По умолчанию после получения сигнала SIGTERM isiot-proxy завершает свою работу не дожидаясь завершения установленных соединений и закрытия сокетов открытых для прослушивания у основного приложения.
+Исправить данное поведение можно путем установки следующей аннотации на pod приложения: `inject.istio.io/templates: sidecar,d8-hold-istio-proxy-termination-until-application-stops`.
+Установка данной аннотаций заставит sidecar-injector вставлять preStop-хук для istio-proxy, который при получении сигнала SIGTERM заставит istio-proxy ждать завершения всех установленных соединений и закрытия сокетов, открытых для прослушивания, но не более чем интервал `terminationGracePeriodSeconds`.
+В некоторых случаях это позволит избежать некорректного завершения приложения.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/USAGE_RU.md:483]:
```suggestion
По умолчанию, в процессе остановки пода, все контейнеры, включая istio-proxy, получают сигнал SIGTERM одновременно. Но некоторым приложениям для правильного завершения работы необходимо время и дополнительная сетевая активность. Это невозможно если istio-proxy завершился раньше.
Решение — добавить в istio-proxy preStop-хук для оценки состояния прикладных сокетов и если таковых нет, то можно останавливать контейнер.
Аннотация ниже добавляет описанный выше preStop-хук в контейнер istio-proxy прикладного пода:
`inject.istio.io/templates: sidecar,d8-hold-istio-proxy-termination-until-application-stops`.
```

**Review by apolovov — APPROVED** [review]

File: ee/modules/110-istio/docs/USAGE_RU.md [hunk 3]
```diff
@@ -473,6 +473,15 @@ spec:
  rules: [{}]
 ```
 
+## Управление поведением data-plane
+
+### [экспериментальная функция] Предотвратить завершение работы istio-proxy до завершения соединений основного приложения
+
+По умолчанию, в процессе остановки пода, все контейнеры, включая istio-proxy, получают сигнал SIGTERM одновременно. Но некоторым приложениям для правильного завершения работы необходимо время и дополнительная сетевая активность. Это невозможно если istio-proxy завершился раньше.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/USAGE_RU.md:480]:
```suggestion
По умолчанию, в процессе остановки пода, все контейнеры, включая istio-proxy, получают сигнал SIGTERM одновременно. Но некоторым приложениям для правильного завершения работы необходимо время и иногда дополнительная сетевая активность. Это невозможно если istio-proxy завершился раньше.
```

File: ee/modules/110-istio/docs/USAGE_RU.md [hunk 4]
```diff
@@ -473,6 +473,15 @@ spec:
...
+Решение — добавить в istio-proxy preStop-хук для оценки состояния прикладных сокетов и если таковых нет, то можно останавливать контейнер.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/USAGE_RU.md:481]:
```suggestion
Решение — добавить в istio-proxy preStop-хук для оценки активности прикладных контейнеров, а единственный доступный метод — это выявление сетевых сокетов приложения, и если таковых нет, то можно останавливать контейнер.
```

File: ee/modules/110-istio/docs/USAGE.md [hunk 5]
```diff
@@ -473,6 +473,14 @@ spec:
  rules: [{}]
 ```
 
+## Control the data-plane behavior
+
+### [experimental feature] Prevent istio-proxy from terminating before the main application's connections are closed
+
+By default, during termination, all containers in a Pod, including istio-proxy one, receive SIGTERM signal simultanuesly. But some applications need time to properly handle the termination and sometimes they need to do some network requests. It isn't possible when the istio-proxy stops before the application do. The solution is to add a preStop hook which evaluates the application's network sockets status and stops when they aren't in network namespace.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/USAGE.md:480]:
```suggestion
By default, during termination, all containers in a Pod, including istio-proxy one, receive SIGTERM signal simultanuesly. But some applications need time to properly handle the termination and sometimes they need to do some network requests. It isn't possible when the istio-proxy stops before the application do. The solution is to add a preStop hook which evaluates the application's activity via discovering application's network sockets and let the sidecar stop when they aren't in network namespace.
```

**Review by apolovov — DISMISSED** [review]

**Review by z9r5 — APPROVED** [review]

**Comment by zbintliff** [comment]
Hello! I see this is very similar to the `holdApplicationUntilProxyStarts` annotation but uses a very different approach to setting it. Does this feature work with `holdApplicationUntilProxyStarts` any reason why it wasn't added as an annotation to `proxy.istio.io/config` like the other? 

 We are getting ready to roll out a change that uses `holdApplicationUntilProxyStarts` but if it is going to change to this format we will hold off for now.

**Comment by fl64** [comment]
@zbintliff Hello!

The `holdApplicationUntilProxyStarts` annotation uses the regular istio functions to implement the postStart hook. While there is no functionality implemented through the `d8-hold-istio-proxy-termination-until-application-stops` template in istio. These annotations do not interfere with each other, and as a result of applying them to the istio-proxy container the following hooks will be added:
```yaml
      lifecycle:
        postStart:
          exec:
            command:
            - pilot-agent
            - wait
        preStop:
          exec:
            command:
            - /bin/sh
            - -c
            - curl -X POST localhost:15000/drain_listeners?inboundonly; while [ $(ss
              -Htlp state all | grep -vE '(envoy|pilot-agent|TIME-WAIT)' | wc -l |
              xargs ) -ne 0 ]; do sleep 1; done
```

**Comment by zbintliff** [comment]
I'm really sorry for my question. I have to admit I found this Issue because it was linked to an issue in Istio. I didn't even realize it went to a different project.

Regardless, thank you for the detailed answer.

**Pull request merged**